src/main/main.ts

```diff
@@ -34,8 +39,87 @@ export function importOnestopToBasecamp(): void {
+        
+        const savedRowBasecampMapping: RowBasecampMapping | null = getRowBasecampMapping(row);
+        
+        if(savedRowBasecampMapping == null) {
+            throw new RowBasecampMappingMissingError("The rowBasecampMapping object is null! Unable to proceed with updating the todo!");
+        }
+        
+        const basecampTodoRequests: Map<string, BasecampTodoRequest> = getBasecampTodoRequestsForRow(row);
+        const currentRoles: string[] = Array.from(basecampTodoRequests.keys());
+
+        const roleTodoIdMap = savedRowBasecampMapping.roleTodoIdMap
+        const originalRoles: string[] = Object.keys(roleTodoIdMap)
+
+        for (const role of currentRoles) {
+
+            let todoId = roleTodoIdMap[role];
+            let request = basecampTodoRequests.get(role);
+
+            if(request == undefined) {
+                throw new BasecampRequestMissingError("Missing basecamp request!");
+                
+            } else if (todoId == undefined) {
+                let newTodoId = createTodo(request, DEFAULT_TODOLIST_IDENTIFIER)
+                roleTodoIdMap[role] = newTodoId // Modify the roleTodoIdMap to reflect changes in the document properties
+                
+            } else {
+                
+                let todoIdentifier: TodoIdentifier = {
+                    projectId: PROJECT_ID,
+                    todoId: todoId
+                }
+                
+                Logger.log(`Updating todo for ${row.what.value} (${row.startTime})...\n`)
+                updateTodo(request, todoIdentifier)
+            }
+        }   
+
+        deleteObsoleteTodos(originalRoles, currentRoles, roleTodoIdMap);
+
+        saveRow(row, roleTodoIdMap);
+    }
+}
+
+/**
+ * Deletes todos associated with helper groups that are no longer present on the event row.
+ * The first line checks whether there are any rows present in the original roles that are not present in the current roles (indicating an obsolete role)
+ * 
+ * @param originalRoles a list of all the original roles associated with the row including the lead role 
+ * @param currentRoles a list of all the current roles associated with the row including the lead role. This may be identical to the original roles
+ */
+function deleteObsoleteTodos(originalRoles: string[], currentRoles: string[], roleTodoIdMap: { [key: string]: string }): void {
+    Logger.log("Checking for obsolete roles...\n")
+    const obsoleteRoles: string[] = originalRoles.filter(role => !currentRoles.includes(role));
+    
+    for(const role of obsoleteRoles) {
+        Logger.log(`Found obsolete role: ${role}!\n`)
+        let todoId = roleTodoIdMap[role];
+
+        let todoIdentifier: TodoIdentifier = {
+            projectId: PROJECT_ID,
+            todoId: todoId
+        }
+
+        Logger.log("Deleting todo...\n")
+        deleteTodo(todoIdentifier);
+
+        delete roleTodoIdMap[role]; // Modify the roleTodoIdMap to reflect changes in the document properties
```

Comment:
as much as possible, we should avoid hidden side-effects with functions (here the roleTodoIdMap is modified)

Reply:
K the PR is updated with this as well